src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+    this.yaxis = null;
+    this.createFramework(fullLayout);
+
+    // state variables used to infer how and what to update
+    this.map = null;
+    this.styleUrl = null;
+    this.traceHash = {};
+    this.layerList = [];
+}
+
+var proto = Mapbox.prototype;
+
+module.exports = function createMapbox(opts) {
+    var mapbox = new Mapbox(opts);
+
+    return mapbox;
+};
+
+proto.plot = function(calcData, fullLayout, promises) {
+    var self = this;
+
+    // feed in new mapbox options
+    self.opts = fullLayout[this.id];
+
+    var promise;
+
+    if(!self.map) {
+        promise = new Promise(function(resolve, reject) {
+            self.createMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+    else {
+        promise = new Promise(function(resolve, reject) {
+            self.updateMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+
+    promises.push(promise);
+};
+
+proto.createMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        gd = self.gd,
+        opts = self.opts;
+
+    // mapbox doesn't have a way to get the current style URL; do it ourselves
+    var styleUrl = self.styleUrl = convertStyleUrl(opts.style);
+
+    var map = self.map = new mapboxgl.Map({
+        container: self.div,
+
+        style: styleUrl,
+        center: convertCenter(opts.center),
+        zoom: opts.zoom,
+        bearing: opts.bearing,
+        pitch: opts.pitch,
+
+        interactive: !self.isStatic,
+        preserveDrawingBuffer: self.isStatic
+    });
+
+    self.rejectOnError(reject);
+
+    map.once('load', function() {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    });
+
+    // keep track of pan / zoom in user layout
+    map.on('move', function() {
+        var center = map.getCenter();
+        opts._input.center = opts.center = { lon: center.lng, lat: center.lat };
+        opts._input.zoom = opts.zoom = map.getZoom();
+    });
+
+    map.on('mousemove', function(evt) {
+        var bb = self.div.getBoundingClientRect();
+
+        // some hackery to get Fx.hover to work
+
+        evt.clientX = evt.point.x + bb.left;
+        evt.clientY = evt.point.y + bb.top;
+
+        evt.target.getBoundingClientRect = function() { return bb; };
+
+        self.xaxis.p2c = function() { return evt.lngLat.lng; };
+        self.yaxis.p2c = function() { return evt.lngLat.lat; };
+
+        Fx.hover(gd, evt, self.id);
+    });
+
+    map.on('click', function() {
+        Fx.click(gd, { target: true });
+    });
+
+    function unhover() {
+        Fx.loneUnhover(fullLayout._toppaper);
+    }
+
+    map.on('dragstart', unhover);
+    map.on('zoomstart', unhover);
+
+};
+
+proto.updateMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        map = self.map;
+
+    self.rejectOnError(reject);
+
+    var styleUrl = convertStyleUrl(self.opts.style);
+
+    if(self.styleUrl !== styleUrl) {
```

Comment:
**Important:** changing map's style is a big relayout step. Mapbox clears a lot of things internally and, in general, we need to redraw all the data to avoid mapbox errors. So, in other words, relayouting `mapbox.style` is a special case.